src/flint/types/fmpz_mod.pyx

```diff
@@ -0,0 +1,428 @@
+                raise NotImplementedError("TODO")
+
+        # Ensure modulus is positive
+        if mod < 1:
+            raise ValueError("Modulus is expected to be positive")
+
+        # Init the context
+        fmpz_mod_ctx_init(self.val, (<fmpz>mod).val)
+    
+    # TODO: should this be cached if we make immutable?
+    def modulus(self):
+        """
+        Return the modulus from the context as an fmpz
+        type
+        """
+        n = fmpz()
+        fmpz_set(n.val, (<fmpz_t>self.val.n))
+        return n
+
+    def __repr__(self):
+        return "Context for fmpz_mod with modulus: {}".format(
+            self.modulus()
+        )
+
+    def __str__(self):
+        return "fmpz_mod_ctx({})".format(
+            self.modulus()
+        )
+
+    def __call__(self, val):
+        if not typecheck(val, fmpz):
+            val = any_as_fmpz(val)
+            if val is NotImplemented:
+                raise NotImplementedError("TODO")
+
+        return fmpz_mod(val, self)
+
+
+cdef class fmpz_mod(flint_scalar):
+    def __init__(self, val, ctx):
+        self.ctx = ctx
+
+        if not typecheck(val, fmpz):
+            val = any_as_fmpz(val)
+            if val is NotImplemented:
+                raise NotImplementedError("TODO")
+
+        fmpz_mod_set_fmpz(self.val, (<fmpz>val).val, self.ctx.val)
+
+    def is_zero(self):
+        return self == 0
+    
+    # TODO: kind of pointless, as we always ensure canonical on init?
+    def is_canonical(self):
+        cdef bint res
+        res = fmpz_mod_is_canonical(self.val, self.ctx.val)
+        return res == 1
+
+    def is_one(self):
+        cdef bint res
+        res = fmpz_mod_is_one(self.val, self.ctx.val)
+        return res == 1
+
+    def __richcmp__(s, t, int op):
+        cdef bint res
+        if op != 2 and op != 3:
+            raise TypeError("fmpz_mod cannot be ordered")
+        # TODO: is this the best method for comparison?
+        if typecheck(s, fmpz_mod) and typecheck(t, fmpz_mod):
+            res = ((<fmpz_mod>s).val[0] == (<fmpz_mod>t).val[0]) and \
+                  ((<fmpz_mod>s).ctx.val.n == (<fmpz_mod>t).ctx.val.n)
+            if op == 2:
+                return res
+            else:
+                return not res
+        # TODO: is this the best method for comparison?
+        # Seems like I'm doing too many type conversions?
+        elif typecheck(s, fmpz_mod) and typecheck(t, int):
+            res = int(s) == t % (<fmpz_mod>s).ctx.modulus()
+            if op == 2:
+                return res
+            else:
+                return not res
+        return NotImplemented
+
+    def __bool__(self):
+        return not self.is_zero()
+
+    def __repr__(self):
+        return "fmpz_mod({}, {})".format(
+            fmpz_get_intlong(self.val),
+            self.ctx.modulus()
+        )
+
+    # TODO: seems ugly...
+    def __hash__(self):
+        return  hash(
+            (int(self), int(self.ctx.modulus()))
+        )
```

Comment:
I think this seems fine but note that `__hash__` should only be defined for immutable types (i.e. no `__iadd__` etc).

Reply:
The hash function here should be consistent with the behaviour of `==` so any time `a == b` is True then we should also have that `hash(a) == hash(b)`.

You allow `a == b` to succeed between `fmpz_mod` and `int` which then means that it should have the same hash as an `int` regardless of modulus.

The hash algorithm used by CPython is:
```
In [24]: e = 10**1000

In [25]: e % (2**61 - 1) == hash(e)
Out[25]: True
```
That is also the same in PyPy. I'm not sure if that should be considered a language feature or not.

The most portable option here is to call `hash(int(self))` but I expect that it can be a lot faster to compute self mod `2**61 - 1` and convert that to an `int`.

Reply:
Some tests should be added to the test suite to cover cases like this to be clear that the python-flint hash matches the int hash.